File: Pinball App 2/Pinball App 2/library/LibraryCatalogVariantLabelSupport.swift
import Foundation

nonisolated func catalogNormalizedVariant(_ value: String?) -> String? {
    guard let trimmed = value?.trimmingCharacters(in: .whitespacesAndNewlines), !trimmed.isEmpty else {
        return nil
    }
    return trimmed
}

nonisolated func catalogResolvedVariantLabel(title: String, explicitVariant: String?) -> String? {
    if let explicitVariant = catalogNormalizedVariantLabel(explicitVariant) {
        return explicitVariant
    }

    let trimmedTitle = title.trimmingCharacters(in: .whitespacesAndNewlines)
    guard trimmedTitle.hasSuffix(")") else { return nil }
    guard let openParenIndex = trimmedTitle.lastIndex(of: "("), openParenIndex > trimmedTitle.startIndex else {
        return nil
    }

    let rawSuffix = trimmedTitle[trimmedTitle.index(after: openParenIndex)..<trimmedTitle.index(before: trimmedTitle.endIndex)]
        .trimmingCharacters(in: .whitespacesAndNewlines)
    guard catalogLooksLikeVariantSuffix(rawSuffix) else { return nil }
    return catalogNormalizedVariantLabel(rawSuffix)
}

nonisolated func catalogResolvedDisplayTitle(title: String, explicitVariant: String?) -> String {
    let trimmedTitle = title.trimmingCharacters(in: .whitespacesAndNewlines)
    guard trimmedTitle.hasSuffix(")") else { return trimmedTitle }
    guard let openParenIndex = trimmedTitle.lastIndex(of: "("), openParenIndex > trimmedTitle.startIndex else {
        return trimmedTitle
    }

    let rawSuffix = trimmedTitle[trimmedTitle.index(after: openParenIndex)..<trimmedTitle.index(before: trimmedTitle.endIndex)]
        .trimmingCharacters(in: .whitespacesAndNewlines)
    guard catalogLooksLikeVariantSuffix(rawSuffix) else { return trimmedTitle }

    let normalizedSuffix = catalogNormalizedVariantLabel(rawSuffix)
    let normalizedExplicit = catalogNormalizedVariantLabel(explicitVariant)
    if let normalizedExplicit, let normalizedSuffix, normalizedExplicit != normalizedSuffix {
        return trimmedTitle
    }

    let baseTitle = trimmedTitle[..<openParenIndex].trimmingCharacters(in: .whitespacesAndNewlines)
    return baseTitle.isEmpty ? trimmedTitle : baseTitle
}

nonisolated func catalogNormalizedVariantLabel(_ value: String?) -> String? {
    guard let trimmed = value?.trimmingCharacters(in: .whitespacesAndNewlines), !trimmed.isEmpty else {
        return nil
    }
    let lowered = trimmed.localizedLowercase
    if lowered == "null" || lowered == "none" {
        return nil
    }
    if lowered == "premium" { return "Premium" }
    if lowered == "pro" { return "Pro" }
    if lowered == "le" || lowered.contains("limited edition") { return "LE" }
    if lowered == "ce" || lowered.contains("collector") { return "CE" }
    if lowered == "se" || lowered.contains("special edition") { return "SE" }
    if lowered == "arcade" { return "Arcade" }
    if lowered == "wizard" { return "Wizard" }
    if lowered == "premium/le" || lowered == "premium le" || lowered == "premium-le" {
        return "Premium/LE"
    }
    if lowered.contains("anniversary") {
        return trimmed
            .split(separator: " ")
            .map { token in
                let loweredToken = token.localizedLowercase
                switch loweredToken {
                case "le", "ce", "se":
                    return loweredToken.uppercased()
                default:
                    return token.prefix(1).uppercased() + token.dropFirst()
                }
            }
            .joined(separator: " ")
    }
    return trimmed
}

private nonisolated func catalogLooksLikeVariantSuffix(_ value: String) -> Bool {
    let lowered = value.localizedLowercase
    return lowered == "premium" ||
        lowered == "pro" ||
        lowered == "le" ||
        lowered == "ce" ||
        lowered == "se" ||
        lowered == "home" ||
        lowered == "arcade" ||
        lowered == "wizard" ||
        lowered.contains("anniversary") ||
        lowered.contains("limited edition") ||
        lowered.contains("special edition") ||
        lowered.contains("collector") ||
        lowered == "premium/le" ||
        lowered == "premium le" ||
        lowered == "premium-le"
}

File: Pinball App Android/app/src/main/java/com/pillyliu/pinprofandroid/library/LibraryCatalogVariantLabelSupport.kt
package com.pillyliu.pinprofandroid.library

internal fun resolvedCatalogVariantLabel(title: String, explicitVariant: String?): String? {
    normalizeCatalogVariantLabel(explicitVariant)?.let { return it }
    val trimmedTitle = title.trim()
    if (!trimmedTitle.endsWith(")")) return null
    val openParenIndex = trimmedTitle.lastIndexOf('(')
    if (openParenIndex <= 0) return null
    val rawSuffix = trimmedTitle.substring(openParenIndex + 1, trimmedTitle.length - 1).trim()
    if (!looksLikeCatalogVariantSuffix(rawSuffix)) return null
    return normalizeCatalogVariantLabel(rawSuffix)
}

internal fun resolvedCatalogDisplayTitle(title: String, explicitVariant: String?): String {
    val trimmedTitle = title.trim()
    if (!trimmedTitle.endsWith(")")) return trimmedTitle
    val openParenIndex = trimmedTitle.lastIndexOf('(')
    if (openParenIndex <= 0) return trimmedTitle
    val rawSuffix = trimmedTitle.substring(openParenIndex + 1, trimmedTitle.length - 1).trim()
    val normalizedSuffix = normalizeCatalogVariantLabel(rawSuffix)
    val normalizedExplicit = normalizeCatalogVariantLabel(explicitVariant)
    if (!looksLikeCatalogVariantSuffix(rawSuffix)) return trimmedTitle
    if (normalizedExplicit != null && normalizedSuffix != null && normalizedExplicit != normalizedSuffix) {
        return trimmedTitle
    }
    return trimmedTitle.substring(0, openParenIndex).trim().ifBlank { trimmedTitle }
}

private fun looksLikeCatalogVariantSuffix(value: String): Boolean {
    val lowered = value.trim().lowercase()
    if (lowered.isBlank()) return false
    return lowered == "premium" ||
        lowered == "pro" ||
        lowered == "le" ||
        lowered == "ce" ||
        lowered == "se" ||
        lowered == "home" ||
        lowered == "arcade" ||
        lowered == "wizard" ||
        lowered.contains("anniversary") ||
        lowered.contains("limited edition") ||
        lowered.contains("special edition") ||
        lowered.contains("collector") ||
        lowered == "premium/le" ||
        lowered == "premium le" ||
        lowered == "premium-le"
}

internal fun normalizeCatalogVariantLabel(value: String?): String? {
    val trimmed = value?.trim().orEmpty()
    if (trimmed.isBlank()) return null
    val lowered = trimmed.lowercase()
    return when {
        lowered == "null" || lowered == "none" -> null
        lowered == "premium" -> "Premium"
        lowered == "pro" -> "Pro"
        lowered == "le" || lowered.contains("limited edition") -> "LE"
        lowered == "ce" || lowered.contains("collector") -> "CE"
        lowered == "se" || lowered.contains("special edition") -> "SE"
        lowered == "arcade" -> "Arcade"
        lowered == "wizard" -> "Wizard"
        lowered == "premium/le" || lowered == "premium le" || lowered == "premium-le" -> "Premium/LE"
        lowered.contains("anniversary") -> trimmed.split(" ")
            .filter { it.isNotBlank() }
            .joinToString(" ") { token ->
                when (token.lowercase()) {
                    "le", "ce", "se" -> token.uppercase()
                    else -> token.replaceFirstChar { ch -> if (ch.isLowerCase()) ch.titlecase() else ch.toString() }
                }
            }
        else -> trimmed
    }
}
